arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptySet.kt

```diff
@@ -0,0 +1,39 @@
+package arrow.core
+
+public class NonEmptySet<out T> private constructor(
+  private val elements: Set<T>
+) : Set<T> by elements {
+
+  public constructor(first: T, rest: Set<T>) : this(setOf(first) + rest)
+
+  public operator fun plus(set: Set<@UnsafeVariance T>): NonEmptySet<T> =
+    NonEmptySet(elements + set)
+
+  public operator fun plus(element: @UnsafeVariance T): NonEmptySet<T> =
+    NonEmptySet(elements + element)
+
+  public fun <R> map(transform: (@UnsafeVariance T) -> R): NonEmptySet<R> =
+    NonEmptySet(elements.map { transform(it) }.toSet())
```

Comment:
How about `elements.mapTo(mutableSetOf(), transform)`

Reply:
Yeap, that's more efficient. Thanks!